File: july2020/Day2_LevelOrder_Traversal2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return 
        
        nodes = [] # result
        queue = deque([root])
        
        # BFS on tree using queue
        
        while(queue):
            level = len(queue)
            levelNodes = []
            
            for i in range(level):
                currentNode = queue.popleft()
                levelNodes.append(currentNode.val)
                
                if currentNode.left:
                    queue.append(currentNode.left)
                if currentNode.right:
                    queue.append(currentNode.right)
                
            nodes.append(levelNodes)
        
        nodes.reverse()
        return nodes

File: july2020/Day2_LevelOrder_Traversal2.java
package july2020;

import java.util.*;

class Solution {
  public List<List<Integer>> levelOrderBottom(TreeNode root) {
      
      ArrayList<List<Integer>> res = new ArrayList<List<Integer>>();
      
      // base case
      if(root == null)
      {
        return res;
      }
      
      ArrayList<Integer> nodes = new ArrayList<Integer>();
      Queue<TreeNode> q = new LinkedList<TreeNode>();
      
      q.offer(root);
      
      while(!q.isEmpty())
      {
          int level = q.size();
          ArrayList<Integer> levelNodes = new ArrayList<Integer>();
          
          for(int i=0; i<level; i++)
          {
              TreeNode currentNode = q.poll();
              
              levelNodes.add(currentNode.val);
              
              if(currentNode.left != null)
                  q.offer(currentNode.left);
              if(currentNode.right != null)
                  q.offer(currentNode.right );        
              
          }
          
          res.add(levelNodes);
          
      }
      
      Collections.reverse(res);
      return res;   
      
  }
}


public class Day2_LevelOrder_Traversal2{

    public static void main(String[] args) {
      Solution obj = new Solution();
      obj.levelOrderBottom(root); 
    }



}